### PR TITLE
Weapons

### PR DESCRIPTION
**Pull request by musique88**
Played around in many files... sorry.

### CODE REVIEW

File: include/World/Player.h [hunk 1]
```diff
@@ -22,11 +23,14 @@ namespace CC::Entities
...
     {
+		//there are clearly better ways to do this
+		std::default_random_engine random_data;
+		bool hasShot;
         CGE::View::Camera &camera_;
         Progression progression;
         Inventory inventory;
 		void init();
-		bool Shoot(World *world);
+		void Shoot(World *world, CC::Weapon &selectedWeapon);
```
Comment (Tursh) [inline include/World/Player.h]: What's going on with the indentations here, lol?

File: include/Inventory/Weapon.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+				WeaponType weaponType = HAND,
+				float precision = 100,
+				float fireRate = 1.f,
+				float damage = 1,
+				float projectileSpeed = 1,
+				float projectileSize = 1.f,
+				int duration = 60,
+				int cost = 0,
+				bool automatic = false);
+		std::shared_ptr<Entities::Projectile> getSharedProjectile() const;
```
Comment (Tursh) [inline include/Inventory/Weapon.h]: In Projectiles.h add "typedef std::shared_ptr<Entities::Projectile> SharedProjectile;" so now you can put "SharedProjectile" as the return type.

File: include/Inventory/Weapon.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+		std::array<int, NUMBER_OF_ELEMENTS> requirements() const;
```
Comment (Tursh) [inline include/Inventory/Weapon.h]: Return a const reference of the array so you don't have to clone the array each time.

File: include/Inventory/Weapon.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+				Element element = NEUTRAL_ELEMENT,
+				WeaponType weaponType = HAND,
+				float precision = 100,
+				float fireRate = 1.f,
+				float damage = 1,
+				float projectileSpeed = 1,
+				float projectileSize = 1.f,
+				int duration = 60,
+				int cost = 0,
+				bool automatic = false);
```
Comment (Tursh) [inline include/Inventory/Weapon.h]: Hmmm... That's a lot of parameters. You could reduce them by giving in the constructor a projectile that contains its speed, size and duration. And btw, the float values are not consistent :p.

File: include/Inventory/Inventory.h [hunk 5]
```diff
@@ -5,17 +5,26 @@
...
 namespace CC
 {
 	class Inventory
 	{
+		std::array<std::optional<Weapon>, 5> toolbar_;
 		std::array<int, NUMBER_OF_ELEMENTS> elemental_;
+
 	public:
+		uint selectedWeapon_;
+		std::array<std::optional<Weapon>, 5> &getToolbar();
```
Comment (Tursh) [inline include/Inventory/Inventory.h]: Return it as a const ref, you don't want anyone to modify the toolbar without passing thought the inventory.

File: include/Inventory/Inventory.h [hunk 6]
```diff
@@ -5,17 +5,26 @@
...
+		void setWeapon(Weapon weapon, int index);
+		Weapon& getWeapon();
```
Comment (Tursh) [inline include/Inventory/Inventory.h]: Same thing here

File: include/World/Player.h [hunk 7]
```diff
@@ -19,7 +23,15 @@ namespace CC::Entities
 
     class Player : public CGE::Entities::Entity
     {
+		//there are clearly better ways to do this
+		std::default_random_engine random_data;
+		bool hasShot;
         CGE::View::Camera &camera_;
+        Progression progression;
+        Inventory inventory;
+		void init();
```
Comment (Tursh) [inline include/World/Player.h]: I'm starting to question if it's github that is showing those weird indentation.

File: src/Inventory/Inventory.cpp [hunk 8]
```diff
@@ -9,11 +9,15 @@ namespace CC
 {
 	Inventory::Inventory(std::array<int,NUMBER_OF_ELEMENTS> elemental)
 	{
+		toolbar_ = {
+				std::make_optional<CC::Weapon>(CC::Weapon()),
+				std::nullopt,std::nullopt,std::nullopt,std::nullopt };
+		selectedWeapon_= 0;
```
Comment (Tursh) [inline src/Inventory/Inventory.cpp]: Use a initializer list to define toolbar_ and selectedWeapon_

File: include/Inventory/Inventory.h [hunk 9]
```diff
@@ -5,17 +5,26 @@
...
 {
 	class Inventory
 	{
+		std::array<std::optional<Weapon>, 5> toolbar_;
 		std::array<int, NUMBER_OF_ELEMENTS> elemental_;
+
 	public:
+		uint selectedWeapon_;
+		std::array<std::optional<Weapon>, 5> &getToolbar();
+		void setWeapon(Weapon weapon, int index);
```
Comment (Tursh) [inline include/Inventory/Inventory.h:21]: You could get the weapon as a const ref if you don't intend to modify it.

File: src/Inventory/Inventory.cpp [hunk 10]
```diff
@@ -31,4 +35,27 @@ namespace CC
...
+	{
+		std::optional<CC::Weapon>& selectedWeapon = getToolbar()[selectedWeapon_];
+		return selectedWeapon ? selectedWeapon.value() : getToolbar()[0].value();
+	}
+
+	void Inventory::update()
+	{
+		for (std::optional<Weapon> &weapon : toolbar_)
+			if (weapon)
+				weapon.value().update();
```
Comment (Tursh) [inline src/Inventory/Inventory.cpp:49]: weapon->update();? I don't know std::optional so not sure.

File: src/Inventory/Projectile.cpp [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+namespace CC::Entities
+{
+	Projectile::Projectile(
+			uint texMeshId, Element element, bool explodes, uint duration, float damage, float size,
+			glm::vec3 position, glm::vec3 rotation,glm::vec3 velocity)
+		:CGE::Entities::Entity(texMeshId, position, rotation),
+		velocity_(velocity), element_(element), explodes_(explodes), duration_(duration), damage_(damage), size_(size)
+	{
+		setSpeed(velocity);
```
Comment (Tursh) [inline src/Inventory/Projectile.cpp:18]: Why you set speed after setting the velocity? SetSpeed => as_ = speed, so just put as_(velocity) in you initializer list.

File: include/Inventory/Projectile.h [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+				bool explodes = false,
+				uint duration = 60,
+				float damage = 1.f,
+				float size = 1.f,
+				glm::vec3 position = glm::vec3(0),
+				glm::vec3 rotation = glm::vec3(0),
+				glm::vec3 velocity = glm::vec3(0));
+		void update() override;
+		//Projectile(const Projectile &projectile);
+		void setPositions(glm::vec3 position = glm::vec3(0),
```
Comment (Tursh) [inline include/Inventory/Projectile.h]: Put all the parameters as const ref.

File: src/World/Player.cpp [hunk 13]
```diff
@@ -27,16 +26,19 @@ namespace CC::Entities
...
+            : Entity(texMeshID, glm::vec3(0, 0, 0.1f), glm::vec3(0.7853f * 2, 0, 0)), camera_(camera) {
+    	init();
+    }
 
     Player::Player(CGE::Loader::TexturedMesh *texMesh, CGE::View::Camera &camera, glm::vec3 position,
                    glm::vec3 rotation)
             : Entity(std::shared_ptr<CGE::Loader::TexturedMesh>(texMesh), position, rotation),
-              camera_(camera) {}
+              camera_(camera) {
+    	init();
```
Comment (Tursh) [inline src/World/Player.cpp:37]: init?

File: src/World/Player.cpp [hunk 14]
```diff
@@ -189,11 +203,33 @@ namespace CC::Entities
...
+    	std::uniform_real_distribution<float> distribution(-1.f, 1.f);
+		std::shared_ptr<CC::Entities::Projectile> projectile = selectedWeapon.getSharedProjectile();
+		projectile->setPositions(
+				getPosition(), getRotation(),
+				camera_.getRotationInNormalizedVector() * selectedWeapon.projectileSpeed_ +
+				glm::vec3(distribution(random_data),distribution(random_data),distribution(random_data))* (1/selectedWeapon.precision_));
+    	world->addEntity(projectile);
+    	selectedWeapon.shoot();
+    	inventory[selectedWeapon.elemental_] -= selectedWeapon.cost_;
+	}
```
Comment (Tursh) [inline src/World/Player.cpp]: You should do those calculations in the weapon function.

**Review by Tursh — CHANGES_REQUESTED** [review]
There are some weird indentation, the shoot function need some modification and the constructor can have less paramters than that

File: include/Inventory/Inventory.h (hunk 6 above)
Comment (musique88) [reply include/Inventory/Inventory.h]: I need to be able to modify the selected weapon for fire rate related calculations

Comment (Tursh) [reply include/Inventory/Inventory.h]: Then do those calculation inside the inventory or the weapon class or don't return it as a ref.

File: src/World/Player.cpp [hunk 15]
```diff
@@ -189,11 +203,30 @@ namespace CC::Entities
...
         glm::vec3 hitBlockPosition = world->getPickedBlock(6.0f);
-
         world->setBlock(hitBlockPosition, Blocks::AIR_BLOCK);
     }
+
+	void Player::init()
+	{
+	}
+
+	void Player::Shoot(World *world, CC::Weapon &selectedWeapon)
```
Comment (Tursh) [inline src/World/Player.cpp]: World could be removed as a parameter here

File: src/World/World.cpp [hunk 16]
```diff
@@ -9,22 +9,27 @@
...
         //TODO create a world generator and check when the player move if chunks can be deleted
-
+		//logInfo(CGE::Utils::TPSClock::getDelta());
         player_->move(0.025f, this);
         for (const auto &entity : entities_)
         {
             entity->update();
+            if (entity->shouldDie())
+			{
+				//TODO Dont know how to destroy shared pointers correctly
```
Comment (Tursh) [inline src/World/World.cpp:28]: We just need a removeEntity in world, but that's a TODO too lol

File: src/World/World.cpp [hunk 17]
```diff
@@ -146,12 +154,12 @@ namespace CC
 
     static glm::vec3 checkCollision(CGE::Entities::Entity *entity, World *world)
     {
-        Physics::Hitbox entityHitbox = entity->getHitbox();
+        Physics::BoxCollider entityHitbox = entity->getHitbox();
```
Comment (Tursh) [inline src/World/World.cpp]: Refactor is tough job

**Review by Tursh — CHANGES_REQUESTED** [review]
Good stuff, but needs some changes

File: src/World/Player.cpp [hunk 18]
```diff
@@ -112,14 +112,14 @@ namespace CC::Entities
 //                hit(world);
 //                lastHit = glfwGetTime();
 //            }
-			CC::Weapon& selectedWeapon = inventory.getWeapon();
+			CC::Weapon selectedWeapon = inventory.getWeapon();
 			if (selectedWeapon.cost_ >= inventory[selectedWeapon.elemental_] && selectedWeapon.canShoot())
```
Comment (Tursh) [inline src/World/Player.cpp]: <=

**Review by Tursh — CHANGES_REQUESTED** [review]
Just a little mistake on your weapon call condition

**Pull request merged**